main-command/src/main/scala/sbt/Watched.scala

```diff
@@ -86,6 +86,11 @@ object Watched {
   def printIfDefined(msg: String) = if (!msg.isEmpty) System.out.println(msg)
 
   def executeContinuously(watched: Watched, s: State, next: String, repeat: String): State = {
+    def isCommandDefined(s: State, commandName: String): Boolean = s.definedCommands.exists {
+      case c: SimpleCommand => c.name == commandName
+      case _                => false
```

Comment:
How would this work for "act" command, a command that represents all of your tasks, like `compile`?
If you run this with `~compile` what do you see?

I am seeing

```
sbt:helloworld> ~compile
[error] Not a valid command: compile (similar: completions)
```

Reply:
Hi,
Good catch!
When I run `~compile` I see exactly the same error message than you.
Do you have any ideas of how I could check as well if the string after the `~` is a task ("act" command).

Reply:
I am not sure if this is possible, but the closest thing that I can think of is how `+` cross building command is implemented here - https://github.com/sbt/sbt/blob/v1.0.2/main/src/main/scala/sbt/Cross.scala#L50-L54. Compare that to how `~` is implemented in https://github.com/sbt/sbt/blob/v1.0.2/main-command/src/main/scala/sbt/BasicCommands.scala#L210-L217 as:

```scala
  def continuous: Command =
    Command(ContinuousExecutePrefix, continuousBriefHelp, continuousDetail)(otherCommandParser) {
      (s, arg) =>
        withAttribute(s, Watched.Configuration, "Continuous execution not configured.") { w =>
          val repeat = ContinuousExecutePrefix + (if (arg.startsWith(" ")) arg else " " + arg)
          Watched.executeContinuously(w, s, arg, repeat)
        }
    }
```

If you want to make `~` more strict, then what you want to do is make `otherCommandParser` more strict. Current implementation of `otherCommandParser` accepts any:

```scala
  lazy val otherCommandParser: State => Parser[String] =
    (s: State) => token(OptSpace ~> combinedLax(s, NotSpaceClass ~ any.*))
```

so you can try replacing it with imple that uses `state.combinedParser` only, but I am not sure if there are other negative implication if you did that, so you'd have to chase the code and figure it out.

Reply:
> but I am not sure if there are other negative implication if you did that, so you'd have to chase the code and figure it out.

Maybe it's the command precedence with other commands. My guess is that `~` uses a relaxed parser intentionally so it can be combined with multi-command (`;`) and cross command (`+`). For example, currently you can write:

```
> ~ ;+compile;show version
```

so sbt is probably saying that for tab completion, I am going to show known commands, but when the user puts in combined commands like the above, it's just going to execute it and see what happens.

Reply:
Any chance you can have a look at this @perbellinioa :smile:?